packages/repository/src/__tests__/unit/repositories/define-repository-class.unit.ts

```diff
@@ -0,0 +1,121 @@
+// Copyright IBM Corp. 2020. All Rights Reserved.
+// Node module: @loopback/repository
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {expect} from '@loopback/testlab';
+import {
+  AndClause,
+  AnyObject,
+  Condition,
+  Count,
+  CrudRepository,
+  DataObject,
+  DefaultCrudRepository,
+  DefaultKeyValueRepository,
+  defineCrudRepositoryClass,
+  defineEntityCrudRepositoryClass,
+  defineKeyValueRepositoryClass,
+  Entity,
+  Filter,
+  model,
+  Model,
+  OrClause,
+  property,
+} from '../../..';
+
+describe('RepositoryClass builder', () => {
+  describe('defineCrudRepositoryClass', () => {
+    it('should generate CRUD repository class', async () => {
+      const AddressRepository = defineCrudRepositoryClass(
+        Address,
+        MockCrudRepository,
+      );
+
+      expect(AddressRepository.name).to.equal('AddressRepository');
+      expect(Object.getPrototypeOf(AddressRepository)).to.equal(
+        MockCrudRepository,
+      );
+    });
+  });
+
+  describe('defineEntityCrudRepositoryClass', () => {
+    it('should generate entity CRUD repository class', async () => {
+      const ProductRepository = defineEntityCrudRepositoryClass(Product);
+
+      expect(ProductRepository.name).to.equal('ProductRepository');
+      expect(Object.getPrototypeOf(ProductRepository)).to.equal(
+        DefaultCrudRepository,
+      );
+    });
+  });
+
+  describe('defineEntityKeyValueRepositoryClass', () => {
+    it('should generate key value repository class', async () => {
+      const ProductRepository = defineKeyValueRepositoryClass(Product);
+
+      expect(ProductRepository.name).to.equal('ProductRepository');
+      expect(Object.getPrototypeOf(ProductRepository)).to.equal(
+        DefaultKeyValueRepository,
+      );
+    });
+  });
+
+  @model()
+  class Product extends Entity {
+    @property({id: true})
+    id: number;
+  }
+
+  @model()
+  class Address extends Model {
+    @property()
+    street: string;
+
+    @property()
+    city: string;
+
+    @property()
+    state: string;
+  }
+
+  class MockCrudRepository<M extends Model> implements CrudRepository<M> {
```

Comment:
There is a subtle difference between mocks and stubs.

Typically, mocks are used when the class implementation is provided entirely by the mocking framework and we want to assert on what mocked methods were called & with what arguments.

```ts
const mock = sinon.mock(myApi);

// run the test invoking mocked methods

sinon.assert.calledOnce(mock.create);
```

Stubs are classes providing a dummy hand-written implementations of the given interface, which is the case of your test file.

Please rename `MockCrudRepository` to `CrudRepositoryStub` or even `DummyCrudRepository`.